compiler/rustc_trait_selection/src/solve/assembly/mod.rs

```diff
@@ -106,10 +109,17 @@ pub(super) trait GoalKind<'tcx>: TypeFoldable<TyCtxt<'tcx>> + Copy + Eq {
         requirements: impl IntoIterator<Item = Goal<'tcx, ty::Predicate<'tcx>>>,
     ) -> QueryResult<'tcx>;
 
-    // Consider a clause specifically for a `dyn Trait` self type. This requires
+    fn consider_alias_bound_clause(
+        ecx: &mut EvalCtxt<'_, 'tcx>,
+        goal: Goal<'tcx, Self>,
+        assumption: ty::Predicate<'tcx>,
+    ) -> QueryResult<'tcx>;
+
+    // FIXME: better wording here.
+    // Consider a clause that is spooky and we cannot trust. This requires
     // additionally checking all of the supertraits and object bounds to hold,
-    // since they're not implied by the well-formedness of the object type.
-    fn consider_object_bound_candidate(
+    // since they're not implied by the well-formedness of anything necessarily.
+    fn consider_non_wf_assumption(
```

Comment:
This is a version of `consider_implied_clause` that registers all of the bounds relevant to a trait, namely its supertraits and item bounds. I assume we'll be using this in all of the built-in impls and for user impls too (or something like it).

Naming bikeshed I guess.